app/models/source_resolver.rb

```diff
@@ -0,0 +1,34 @@
+class SourceResolver
+  def self.resolve(new_source)
+    new(new_source).resolve
+  end
+
+  def initialize(new_source)
+    @new_source = new_source
+  end
+
+  def resolve
+    return if @new_source.save
+    adjust_ranks
```

Comment:
This is interesting. You're probably right that a transaction does improve the safety of this code, but I'm having a hard time writing a failing test that proves the need. Open to ideas. Or I could just add the transaction, but it feels like anticipating a problem rather than responding to one, ya know?